### PR TITLE
[UPLOAD-731] keycloak

### PR DESCRIPTION
**Pull request by krystophv**
For [UPLOAD-731], adds keycloak login support. Largely an analog of [WEB-1372].
In addition, adds support for remembering chosen environment on launch.

[UPLOAD-731]: https://tidepool.atlassian.net/browse/UPLOAD-731?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[WEB-1372]: https://tidepool.atlassian.net/browse/WEB-1372?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by clintonium-119 — APPROVED** [review]
Excellent work :muscle: 
Couple of optional removal of comments, but very cleanly done :+1:

File: electron-builder-publish.js [hunk 1]
```diff
@@ -104,7 +108,7 @@ if ( (process.env.CIRCLE_TAG && process.env.CIRCLE_TAG.length > 0) ||
   if ( (process.env.CIRCLE_TAG && process.env.CIRCLE_TAG.indexOf('-') !== -1) ||
        (process.env.APPVEYOR_REPO_TAG_NAME && process.env.APPVEYOR_REPO_TAG_NAME.indexOf('-') !== -1) ) {
     // non-production releases have hyphens in their tags
```
[the remark this answers is not in the document]
Comment (krystophv) [reply electron-builder-publish.js:110]: Yeah. I didn't add the comment and it seemed valid, if a bit obvious/redundant as to what was going on so I didn't think to remove it initially. It does seem redundant enough that it's not terribly useful though.

File: lib/core/api.js [hunk 2]
```diff
@@ -148,6 +154,23 @@ api.user.loginExtended = (user, options, cb) => {
...
+api.user.saveSession = (userId, token, options, cb) => {
+  options = options || {};
+  if (typeof options === 'function') {
+    cb = options;
+    options = {};
+  }
+  if (_.isUndefined(cb)) {
+    cb = _.noop;
+  }
+  tidepool.saveSession(userId, token, options, (err /* , session */) => {
```
[the remark this answers is not in the document]
Comment (krystophv) [reply lib/core/api.js:166]: This was more for me just documenting the API for myself in development. While it's _potentially_ useful to have the callback function argument signature available for the future, it's probably not worth just leaving this hanging around. 👍🏼

**Review by clintonium-119 — APPROVED** [review]
LGTM 🎸

**Review by clintonium-119 — DISMISSED** [review]
Changes LGTM :rocket:

**Review by clintonium-119 — APPROVED** [review]
LGTM :shipit:

**Pull request merged**